subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CompilationClassBackupService.java

```diff
@@ -0,0 +1,86 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.Set;
+
+/**
+ * A service that backups a class just before new version of a class is generated.
+ * This is used for restoring compiler outputs on compilation failure.
+ *
+ * Service only backups classes that:
+ * - exists before compilation
+ * - were not deleted before compilation
+ * - javac tries to overwrite during compilation.
+ *
+ * This service is called by compiler through Compiler API.
+ *
+ * We don't backup classes/resources generated by annotation processors, since any overwriting of existing generated types
+ * that were not deleted by Gradle before annotation processing, would anyway result in full-recompilation, due to annotation processors limitations.
+ * And in case of a Groovy/Java joint compilation, incremental compilation is anyway disabled, when annotation processing is present.
+ */
+public class CompilationClassBackupService {
+
+    private final Set<String> classesToCompile;
+    private final File destinationDir;
+    private final File headerOutputDir;
+    private final File classBackupDir;
+    private final ApiCompilerResult result;
+
+    public CompilationClassBackupService(JavaCompileSpec spec, ApiCompilerResult result) {
+        this.classesToCompile = spec.getClassesToCompile();
+        this.destinationDir = spec.getDestinationDir();
+        this.headerOutputDir = spec.getCompileOptions().getHeaderOutputDirectory();
+        this.classBackupDir = spec.getClassBackupDir();
+        this.result = result;
+    }
+
+    public void maybeBackupClassFile(String classFqName) {
+        // Classes to compile are stashed before the compilation, so there is nothing to backup
+        if (classBackupDir != null && !classesToCompile.contains(classFqName)) {
+            String classFilePath = classFqName.replace(".", "/").concat(".class");
+            maybeBackupFile(classFqName, destinationDir, classFilePath, ".class");
+            if (headerOutputDir != null) {
+                String headerFilePath = classFqName.replaceAll("[.$]", "_").concat(".h");
+                maybeBackupFile(classFqName, headerOutputDir, headerFilePath, ".h");
+            }
+        }
+    }
+
+    private void maybeBackupFile(String classFqName, File destinationDir, String relativePath, String suffix) {
+        File classFile = new File(destinationDir, relativePath);
+        if (!result.getBackupClassFiles().containsKey(classFile.getAbsolutePath()) && classFile.exists()) {
+            File backupFile = new File(classBackupDir, classFqName + suffix);
```

Comment:
🤔 Would it make sense to use a unique id here for the backup file instead of the FQCN?

Reply:
Added unique id

Reply:
We can also do that. I guess this would make it a bit faster when serializing/deserializing worker result although probably this would be small, since for normal case we would not backup any file. Or do you see any other benefit?

Reply:
So something like `class-<index>.backup` or `class-backup-<index>.class`?

Reply:
The reason would be that the name doesn't become too long, which may be a problem on Windows. Before you used `sourceFile.getName() + ".uniqueId"`, I think it would make sense to do the same again.